### PR TITLE
Make sequence a property

### LINKED ISSUE #271: Hidden properties should be replaced with Python property decorators
The current mechanism for aliasing properties and hiding them from serialization (libSBOL "hidden" properties) can be better implemented using Python property decorators

### LINKED ISSUE #272: Objects contained in a hidden property shouldn't persist if they are  removed from the Document top level
The following use case illustrates the problem:
```
doc = sbol2.Document()
cd = doc.componentDefinitions.create('cd1')
cd.sequence = sbol2.Sequence('cd1_sequence')
self.assertIn('cd1_sequence', doc.sequences)
doc.sequences.remove('cd1_sequence')
self.assertIsNone(cd.sequence)
```
This assertion fails, because the object in `cd.sequence` persists even though it was removed from the `Document` top level.

One way to address this is to call `find` from inside the `remove` method, but this proves to be a computationally expensive operation.

Alternatively, rethinking the current approach for "hidden" properties (https://github.com/SynBioDex/pySBOL2/issues/271) may resolve this issue in a more elegant and efficient manner

### PR DESCRIPTION
**Pull request by tcmitchell**
Change sequence from a hidden property to a Python property.

Fixes #271 
Fixes #272

### CODE REVIEW

File: sbol2/componentdefinition.py [hunk 1]
```diff
@@ -134,7 +129,46 @@ def __init__(self, uri=URIRef("example"),
...
+        if self.doc:
+            try:
+                return self.doc.find(seq_uri)
+            except NotImplementedError:
+                raise
+
+    @sequence.setter
+    def sequence(self, sequence: Union[Sequence, None]):
+        if not sequence:
+            # TODO: How to remove an orphaned sequence from the document?
```
Comment (bbartley) [inline sbol2/componentdefinition.py:154]: I don't think this is an issue.  Even if the Sequence is set to None, it's still OK for that Sequence to persist in the Document (e.g., maybe the user wants to re-use the Sequence with some other ComponentDefinition)

File: sbol2/componentdefinition.py [hunk 2]
```diff
@@ -134,7 +129,46 @@ def __init__(self, uri=URIRef("example"),
...
+        if self.doc:
+            try:
+                self.doc.add(sequence)
+            except SBOLError as e:
+                if e.error_code() != SBOLErrorCode.DUPLICATE_URI_ERROR:
+                    raise
+        self._sequence_cache = sequence
+        self.sequences = [sequence.identity]
+
+    def added_to_document(self, doc):
```
Comment (bbartley) [inline sbol2/componentdefinition.py:167]: Maybe make this private?  (In the super as well)

File: sbol2/componentdefinition.py [hunk 3]
```diff
@@ -134,7 +129,46 @@ def __init__(self, uri=URIRef("example"),
...
+        super().added_to_document(doc)
+        # Add the sequence to the document
+        if self._sequence_cache:
+            doc.add(self._sequence_cache)
```
Comment (bbartley) [inline sbol2/componentdefinition.py:171]: Do we need a check for removal as well?

Comment (tcmitchell) [reply sbol2/componentdefinition.py:171]: I was wondering the same thing myself. This is how I finessed adding the `Sequence` to the `Document` when the `ComponentDefinition` is added to the `Document`. We have a unit test for that.

I'm not sure that makes sense in a pure SBOL world, but I'd bet there are programs out there that rely on that capability.

File: sbol2/componentdefinition.py (hunk 2 above)
Comment (tcmitchell) [reply sbol2/componentdefinition.py:167]: I don't want to make it private because it is called by `Document`. I think it should be part of the public API of `SBOLObject` instances. We can discuss it and see what we think.

File: sbol2/componentdefinition.py [hunk 4]
```diff
@@ -134,7 +129,46 @@ def __init__(self, uri=URIRef("example"),
...
+            return None
+        seq_uri = seqs[0]
+        if self._sequence_cache and self._sequence_cache.identity == seq_uri:
+            return self._sequence_cache
+        # Not in the cache, try to look up in the document
+        if self.doc:
+            try:
+                return self.doc.find(seq_uri)
+            except NotImplementedError:
+                raise
```
[the remark this answers is not in the document]
Comment (tcmitchell) [reply sbol2/componentdefinition.py]: I revamped the code. `_sequence_cache` is only used if the ComponentDefinition is _not_ in a Document. If the CD is in a Document, then `_sequence_cache` is not used and a lookup is done instead.

File: sbol2/componentdefinition.py (hunk 3 above)
Comment (tcmitchell) [reply sbol2/componentdefinition.py:180]: I'm reluctant to add a `SBOLObject.removed_from_document` method because we have no use for it right now.

File: sbol2/componentdefinition.py (hunk 2 above)
Comment (bbartley) [reply sbol2/componentdefinition.py:176]: I'm thinking it should be private because a casual user will never need to call it in the context of a script.  So it clutters the public API.  Even an advanced user would only need it if they are defining custom class behaviors, which probably means only us.

Since we can't make it private, is there any other way to hide it?  If there's nothing straightforward, then we should just leave it as is.

File: sbol2/componentdefinition.py (hunk 3 above)
Comment (bbartley) [reply sbol2/componentdefinition.py:180]: The edge case I'm thinking about is `test_remove_hidden_sequence` in `test_componentdefinition.py`.  Does this test pass given your changes above?  If not, then there's probably some extra handling that can go here to deal with it.

Comment (bbartley) [reply sbol2/componentdefinition.py:180]: Use a friend class!  (J/k, that's just a little C++ humor to brighten your day)

Comment (tcmitchell) [reply sbol2/componentdefinition.py:180]: `test_remove_hidden_sequence` in `test_componentdefinition.py` does pass with these changes. I removed the `@unittest.expectedFailure` as part of these changes.

I think what makes this work is the lookup via the document, which ignores a cached sequence. So once the Sequence is removed from the document, the ComponentDefinition can no longer look it up.

File: sbol2/componentdefinition.py (hunk 2 above)
Comment (tcmitchell) [reply sbol2/componentdefinition.py]: I'm convinced. I changed the method name to `_added_to_document`.

File: sbol2/componentdefinition.py (hunk 3 above)
Comment (bbartley) [reply sbol2/componentdefinition.py:180]: Cool.  That's elegant.

File: sbol2/componentdefinition.py (hunk 2 above)
Comment (bbartley) [reply sbol2/componentdefinition.py]: Nice
